File: R/04_CNN_real_image with kaggle cats & dogs.R
library(keras)

data_from_dir='C:/Users/young/Desktop/R/cats-dogs'
base_dir='C:/Users/young/Desktop/R/kaggle_cat'
dir.create(base_dir)


#dir.create
train_dir=file.path(base_dir,'train')
val_dir=file.path(base_dir,'val')
test_dir=file.path(base_dir,'test')

train_cats_dir=file.path(train_dir,'cats')
train_dogs_dir=file.path(train_dir,'dogs')
val_cats_dir=file.path(val_dir,'cats')
val_dogs_dir=file.path(val_dir,'dogs')
test_cats_dir=file.path(test_dir,'cats')
test_dogs_dir=file.path(test_dir,'dogs')

# copy file
fnames=paste0('cat.',1:1000,'.jpg')
file.copy(file.path(data_from_dir,fnames),
          file.path(train_cats_dir))
fnames=paste0('dog.',1:1000,'.jpg')
file.copy(file.path(data_from_dir,fnames),
          file.path(train_dogs_dir))
fnames=paste0('cat.',1001:1500,'.jpg')
file.copy(file.path(data_from_dir,fnames),
          file.path(val_cats_dir))
fnames=paste0('dog.',1001:1500,'.jpg')
file.copy(file.path(data_from_dir,fnames),
          file.path(val_dogs_dir))
fnames=paste0('cat.',1501:2000,'.jpg')
file.copy(file.path(data_from_dir,fnames),
          file.path(test_cats_dir))
fnames=paste0('dog.',1501:2000,'.jpg')
file.copy(file.path(data_from_dir,fnames),
          file.path(test_dogs_dir))

# make generator
data_gen=image_data_generator(rescale = 1/255)

train_gen=flow_images_from_directory(
  train_dir,
  data_gen,
  target_size = c(150,150),
  batch_size = 20,
  class_mode = 'binary'
)

val_gen=flow_images_from_directory(
  val_dir,
  data_gen,
  target_size = c(150,150),
  batch_size = 20,
  class_mode = 'binary'
)

# build model
model=keras_model_sequential()

model %>% 
  layer_conv_2d(filters = 32,activation='relu',input_shape = c(150,150,3),kernel_size = c(3,3)) %>% 
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 
  layer_conv_2d(filter=64,activation='relu',kernel_size = c(3,3)) %>% 
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 
  layer_conv_2d(filter=64,activation='relu',kernel_size = c(3,3)) %>% 
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 
  layer_conv_2d(filter=128,activation='relu',kernel_size = c(3,3)) %>% 
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 
  layer_flatten() %>% 
  layer_dense(units=512,activation='relu') %>% 
  layer_dense(units = 1,activation='sigmoid')

model

model %>% compile(
  optimizer='rmsprop',
  loss='binary_crossentropy',
  metrics=c('acc')
)

# run/save/load model
history=model %>% fit_generator(
  train_gen,
  steps_per_epoch = 100,
  epochs = 10,
  validation_data = val_gen,
  validation_steps = 50
)

model %>% save_model_hdf5('cats_dogs.h5')
# m = load_model_hdf5('C:/Users/young/Desktop/R/cats_dogs.h5')

File: python/04_CNN_real_image with kaggle cats & dogs.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import pandas as pd
import numpy as np
import keras
import matplotlib.pyplot as plt


# In[2]:


data_from_dir = 'C:/Users/young/Desktop/R/cats-dogs'
base_dir = base_dir='C:/Users/young/Desktop/R/kaggle_cat'

train_dir='C:/Users/young/Desktop/R/kaggle_cat/train'
val_dir='C:/Users/young/Desktop/R/kaggle_cat/val'
test_dir='C:/Users/young/Desktop/R/kaggle_cat/test'

train_cats_dir=f'C:/Users/young/Desktop/R/kaggle_cat/train/cats'
train_dogs_dir='C:/Users/young/Desktop/R/kaggle_cat/train/dogs'
val_cats_dir='C:/Users/young/Desktop/R/kaggle_cat/val/cats'
val_dogs_dir='C:/Users/young/Desktop/R/kaggle_cat/val/dogs'
test_cats_dir='C:/Users/young/Desktop/R/kaggle_cat/test/cats'
test_dogs_dir='C:/Users/young/Desktop/R/kaggle_cat/test/dogs'


# In[3]:


from keras.preprocessing.image import ImageDataGenerator

train_data_gen = ImageDataGenerator(rescale = 1/255,
                             zoom_range= 0.2)
data_gen = ImageDataGenerator(rescale = 1/255)

train_gen = train_data_gen.flow_from_directory(train_dir,
                                        target_size=(150,150),
                                        batch_size=20,
                                        class_mode='binary')
val_gen = data_gen.flow_from_directory(val_dir,
                                        target_size=(150,150),
                                        batch_size=20,
                                        class_mode='binary')
test_gen = data_gen.flow_from_directory(test_dir,
                                        target_size=(150,150),
                                        batch_size=20,
                                        class_mode='binary')


# In[4]:


from keras import models, layers

model = models.Sequential()

model.add(layers.Conv2D(32, activation='relu', input_shape=(150,150,3), kernel_size=3))
model.add(layers.MaxPool2D(2))
model.add(layers.Conv2D(32, activation='relu', kernel_size=3))
model.add(layers.MaxPool2D(2))
model.add(layers.Conv2D(32, activation='relu',  kernel_size=3))
model.add(layers.MaxPool2D(2))
model.add(layers.Conv2D(32, activation='relu',  kernel_size=3))
model.add(layers.MaxPool2D(2))
model.add(layers.Flatten())
model.add(layers.Dense(256, activation = 'relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.summary()

model.compile(optimizer='rmsprop', 
              loss='binary_crossentropy', 
              metrics=['acc'])


# In[5]:


history = model.fit_generator(train_gen,
                             steps_per_epoch=100,
                             epochs = 10,
                             validation_data= val_gen,
                             validation_steps=50)


# In[6]:


plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('Model loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='upper right')
plt.show()


# In[7]:


plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('Model accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='upper left')
plt.show()


# In[8]:


from keras import preprocessing


# In[9]:


img = preprocessing.image.load_img('C:/Users/young/Desktop/R/kaggle_cat/test/cats/cat.1601.jpg', 
                               target_size = (150,150))
img = preprocessing.image.img_to_array(img)
img = np.reshape(img, (1,150,150,3))/255


# In[10]:


from keras.models import Model

layer_outputs = [layer.output for layer in model.layers[:8]]
activation_model = keras.models.Model(inputs = model.input, outputs=layer_outputs)

activations = activation_model.predict(img)


# In[11]:


plt.matshow(activations[0][0,:,:,1])
plt.matshow(activations[0][0,:,:,7])
plt.matshow(activations[0][0,:,:,10])
plt.matshow(activations[2][0,:,:,1])
plt.matshow(activations[2][0,:,:,7])
plt.matshow(activations[2][0,:,:,10])


# In[ ]:




